wagtail/admin/views/pages/choose_parent.py

```diff
@@ -18,6 +19,67 @@ class ChooseParentView(WagtailAdminTemplateMixin, FormView):
     index_url_name = None
     page_title = gettext_lazy("Choose parent")
 
+    def get_valid_parent_pages(self, user):
+        """
+        Identifies possible parent pages for the current user by first looking
+        at allowed_parent_page_models() on self.model to limit options to the
+        correct type of page, then checking permissions on those individual
+        pages to make sure we have permission to add a subpage to it.
+        """
+        # Get queryset of pages where this page type can be added
+        allowed_parent_page_content_types = list(
+            ContentType.objects.get_for_models(
+                *self.model.allowed_parent_page_models()
```

Comment:
For the test, you can probably get away with mocking this method so you don't have to set up the viewset etc., e.g. something like

```py
    def test_skip_if_only_one_valid_parent(self):
        self.assertEqual(EventIndex.objects.count(), 1)

        with mock.patch.object(EventPage, "allowed_parent_page_models") as mock_method:
            mock_method.return_value = [EventIndex]
            self.assertEqual(EventPage.allowed_parent_page_models(), [EventIndex])
            response = self.client.get(self.view_url)

        self.assertRedirects(
            response,
            reverse(
                "wagtailadmin_pages:add",
                args=("tests", "eventpage", EventIndex.objects.first().pk),
            ),
        )
```

in https://github.com/wagtail/wagtail/blob/main/wagtail/admin/tests/pages/test_parent_page_chooser_view.py

Reply:
@laymonage thanks sage! I would have never thought to do it that way :) I see how that makes it easier and tests the redirect. But this way we have not tested the `allowed_parent_page_models` function to correctly return one parent page if only one is available and user has the permissions, right? Why is it ok to skip that part?

Reply:
That's a good question! I think it's okay because the `allowed_parent_page_models` method already has its own test here:
https://github.com/wagtail/wagtail/blob/06ad1d9dc122b64cad9a0610bd768377bd3a38cd/wagtail/tests/test_page_model.py#L3048-L3063

and the method's logic isn't tied to the user's permissions, so mocking this method is essentially the same as setting `EventPage.parent_page_types` directly (or setting up another `PageListingViewSet` with a model that has the attribute set).

Just to be clear, we are not mocking the view's `get_valid_parent_pages` here. We're mocking the page model's method, and the rest of the view's logic still runs.

The test case example I gave above only test for the superuser case. For the other case, I think we can even skip the mock and just give the user a permission to a leaf page. For example:

```py
    def test_user_with_only_one_valid_parent_page(self):
        test_group = Group.objects.create(name="test_group")
        test_group.permissions.add(Permission.objects.get(codename="access_admin"))

        self.user.is_superuser = False
        self.user.groups.add(test_group)
        self.user.save()

        # Only grant the user permission to add pages under one leaf page
        parent_page = SimplePage.objects.filter(numchild=0).first()
        GroupPagePermission.objects.create(
            group=test_group,
            page=parent_page,
            permission_type="add",
        )

        # Should redirect to the add view with the only valid parent page
        response = self.client.get(self.view_url)
        self.assertRedirects(
            response,
            reverse(
                "wagtailadmin_pages:add",
                args=("tests", "eventpage", parent_page.pk),
            ),
        )
```

Reply:
> Just to be clear, we are not mocking the view's `get_valid_parent_pages` here. We're mocking the page model's method, and the rest of the view's logic still runs.

Ah yes you are right. Sorry i was confused and thought we had mocked the `get_valid_parent_pages`. Makes total sense! Thanks for the clarification.